File: 0328-odd-even-linked-list/0328-odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        
        if (head == null || head.next == null || head.next.next == null) {
            return head;
        }
        
        ListNode odd = head;
        
        ListNode evenh = head.next;
        ListNode even = evenh;
        
        ListNode ptr = head.next.next;
        
        while (ptr != null) {
            odd.next = ptr;
            odd = odd.next;
            
            even.next = ptr.next;
            even = even.next;
            
            if (ptr.next == null) {
                break;
            }
            ptr = ptr.next.next;
            
        }
        
        odd.next = evenh;
        return head;
        
    }
}

File: 0328-odd-even-linked-list/0328-odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        if head is None or head.next == None or head.next.next == None:
            return head
        
        evenHead = head.next
        
        oddCurrent = head
        evenCurrent = evenHead
        
        while oddCurrent != None or evenCurrent != None:
            
            oddCurrent.next = evenCurrent.next
            evenCurrent.next = evenCurrent.next.next
            
            oddCurrent = oddCurrent.next
            if evenCurrent.next != None:
                evenCurrent = evenCurrent.next
            
            if evenCurrent.next == None:
                break
        
        oddCurrent.next = evenHead
        
        return head
        